# Объекты даты и времени
 - datetime
 - time
 - calendar
 
 2 типа объектов:
 - native
 - aware

For applications requiring aware objects, datetime and time objects have an optional time zone information attribute, tzinfo, that can be set to an instance of a subclass of the abstract tzinfo class. 

In [1]:
import datetime

In [2]:
datetime.MINYEAR

1

In [3]:
datetime.MAXYEAR

9999

In [19]:
datetime.datetime.now().date()

datetime.date(2019, 3, 23)

In [8]:
d = datetime.datetime(2019, 3, 23, 13, 10, 52, 99530)

In [9]:
type(d)

datetime.datetime

In [10]:
dir(datetime)

['MAXYEAR',
 'MINYEAR',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'date',
 'datetime',
 'datetime_CAPI',
 'sys',
 'time',
 'timedelta',
 'timezone',
 'tzinfo']

In [5]:
datetime.date.today()

datetime.date(2019, 3, 23)

## Date
```python
datetime.date # class
```

In [12]:
date = datetime.date(2017, 1, 15)
datetime.date(

In [13]:
type(date)

datetime.date

## Time
```python
datetime.time # class
```

In [15]:
time = datetime.time(23, 59, 59)
datetime.time(

In [17]:
print(time) # no tzinfo

23:59:59


In [20]:
print(time.tzinfo)

None


In [21]:
import pytz

In [22]:
pytz.all_timezones

['Africa/Abidjan', 'Africa/Accra', 'Africa/Addis_Ababa', 'Africa/Algiers', 'Africa/Asmara', 'Africa/Asmera', 'Africa/Bamako', 'Africa/Bangui', 'Africa/Banjul', 'Africa/Bissau', 'Africa/Blantyre', 'Africa/Brazzaville', 'Africa/Bujumbura', 'Africa/Cairo', 'Africa/Casablanca', 'Africa/Ceuta', 'Africa/Conakry', 'Africa/Dakar', 'Africa/Dar_es_Salaam', 'Africa/Djibouti', 'Africa/Douala', 'Africa/El_Aaiun', 'Africa/Freetown', 'Africa/Gaborone', 'Africa/Harare', 'Africa/Johannesburg', 'Africa/Juba', 'Africa/Kampala', 'Africa/Khartoum', 'Africa/Kigali', 'Africa/Kinshasa', 'Africa/Lagos', 'Africa/Libreville', 'Africa/Lome', 'Africa/Luanda', 'Africa/Lubumbashi', 'Africa/Lusaka', 'Africa/Malabo', 'Africa/Maputo', 'Africa/Maseru', 'Africa/Mbabane', 'Africa/Mogadishu', 'Africa/Monrovia', 'Africa/Nairobi', 'Africa/Ndjamena', 'Africa/Niamey', 'Africa/Nouakchott', 'Africa/Ouagadougou', 'Africa/Porto-Novo', 'Africa/Sao_Tome', 'Africa/Timbuktu', 'Africa/Tripoli', 'Africa/Tunis', 'Africa/Windhoek', 'Ameri

In [23]:
timezone = pytz.timezone("America/Belem")

In [24]:
timezone

<DstTzInfo 'America/Belem' LMT-1 day, 20:46:00 STD>

In [30]:
dir(pytz)

['AmbiguousTimeError',
 'BaseTzInfo',
 'FixedOffset',
 'HOUR',
 'InvalidTimeError',
 'LazyDict',
 'LazyList',
 'LazySet',
 'NonExistentTimeError',
 'OLSEN_VERSION',
 'OLSON_VERSION',
 'UTC',
 'UnknownTimeZoneError',
 'VERSION',
 'ZERO',
 '_CountryNameDict',
 '_CountryTimezoneDict',
 '_FixedOffset',
 '_UTC',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_p',
 '_test',
 '_tzinfo_cache',
 '_unmunge_zone',
 'all_timezones',
 'all_timezones_set',
 'ascii',
 'build_tzinfo',
 'common_timezones',
 'common_timezones_set',
 'country_names',
 'country_timezones',
 'datetime',
 'exceptions',
 'lazy',
 'open_resource',
 'os',
 'resource_exists',
 'sys',
 'timezone',
 'tzfile',
 'tzinfo',
 'unicode',
 'unpickler',
 'utc']

In [25]:
time = datetime.time(23, 59, 59, tzinfo=timezone)

In [26]:
timezone1 = pytz.timezone("Europe/Minsk")

In [27]:
time1 = datetime.time(23, 59, 59, tzinfo=timezone1)

In [32]:
dir(time)

['__class__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'dst',
 'fold',
 'fromisoformat',
 'hour',
 'isoformat',
 'max',
 'microsecond',
 'min',
 'minute',
 'replace',
 'resolution',
 'second',
 'strftime',
 'tzinfo',
 'tzname',
 'utcoffset']

In [36]:
print(time.utcoffset())

None


In [37]:
rb = datetime.date(1991, 8, 25)
today = datetime.date.today()

In [42]:
diff = today - rb

In [43]:
diff

datetime.timedelta(days=10072)

In [52]:
td = datetime.timedelta(10000)

In [53]:
td

datetime.timedelta(days=10000)

In [57]:
print(rb - today)

-10072 days, 0:00:00


In [44]:
dir(datetime.timedelta)

['__abs__',
 '__add__',
 '__bool__',
 '__class__',
 '__delattr__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__pos__',
 '__radd__',
 '__rdivmod__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rfloordiv__',
 '__rmod__',
 '__rmul__',
 '__rsub__',
 '__rtruediv__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__truediv__',
 'days',
 'max',
 'microseconds',
 'min',
 'resolution',
 'seconds',
 'total_seconds']

In [40]:
rb + 3

TypeError: unsupported operand type(s) for +: 'datetime.date' and 'int'

In [63]:
print( f'Дней: {diff.total_seconds()}')

Дней: 870220800.0


### Преобразования из/в строку

In [64]:
tn = datetime.datetime.now()

In [65]:
type(tn)

datetime.datetime

In [74]:
str_tn = tn.strftime("%d.%m.%y %H:%M %B") # time to string

In [75]:
str_tn

'23.03.19 13:39 March'

In [83]:
time_as_string = '18.03.19 13:27'

In [77]:
time_obj = datetime.datetime.strptime??

In [ ]:
time_obj = datetime.datetime.strptime

In [ ]:
time_obj = datetime.datetime.strptime

In [79]:
time_obj1 = datetime.datetime(day=1, month=1, year=2001)

In [82]:
time_as_string

'18.03.2019 13:27'

In [85]:
time_obj = datetime.datetime.strptime(time_as_string, "%d.%m.%y %H:%M")

In [86]:
time_obj

datetime.datetime(2019, 3, 18, 13, 27)

In [ ]:
time_obj

In [ ]:
tn


| Directive | Meaning |	Example	| Notes |
| :--- | :--- | :--- | :--- |
| %a |	Weekday as locale’s abbreviated name. | Sun, Mon, …, Sat (en_US); So, Mo, …, Sa (de_DE) | (1) |
 |%A |	Weekday as locale’s full name. | Sunday, Monday, …, Saturday (en_US); Sonntag, Montag, …, Samstag (de_DE) | (1) |
|%w	|Weekday as a decimal number, where 0 is Sunday and 6 is Saturday.	|0, 1, …, 6| |
|%d	|Day of the month as a zero-padded decimal number.	|01, 02, …, 31	| | 
|%b	|Month as locale’s abbreviated name.	|Jan, Feb, …, Dec (en_US); Jan, Feb, …, Dez (de_DE) | (1)|
|%B	|Month as locale’s full name.	| January, February, …, December (en_US); Januar, Februar, …, Dezember (de_DE) | (1)|
|%m	|Month as a zero-padded decimal number.|	01, 02, …, 12| |	 
|%y	|Year without century as a zero-padded decimal number.	|00, 01, …, 99	| | 
|%Y	|Year with century as a decimal number.	|0001, 0002, …, 2013, 2014, …, 9998, 9999|	(2)|
|%H	|Hour (24-hour clock) as a zero-padded decimal number.	|00, 01, …, 23	 | |
|%I	|Hour (12-hour clock) as a zero-padded decimal number.	|01, 02, …, 12	 | |
|%p	|Locale’s equivalent of either AM or PM. |AM, PM (en_US); am, pm (de_DE)| (1), (3)
|%M	|Minute as a zero-padded decimal number.	|00, 01, …, 59	| | 
|%S	|Second as a zero-padded decimal number.	|00, 01, …, 59	|(4)|
|%f	|Microsecond as a decimal number, zero-padded on the left.	|000000, 000001, …, 999999| (5)|
|%z	|UTC offset in the form ±HHMM[SS[.ffffff]] (empty string if the object is naive).|(empty), +0000, -0400, +1030, +063415, -030712.345216	|(6)|
|%Z	|Time zone name (empty string if the object is naive).|	(empty), UTC, EST, CST| |	 
|%j	|Day of the year as a zero-padded decimal number.	|001, 002, …, 366| |	 
|%U	|Week number of the year (Sunday as the first day of the week) as a zero padded decimal number. All days in a new year preceding the first Sunday are considered to be in week 0.	|0, 01, …, 53|	(7)|
|%W	|Week number of the year (Monday as the first day of the week) as a decimal number. All days in a new year preceding the first Monday are considered to be in week 0.|	00, 01, …, 53|(7)|
|%c	|Locale’s appropriate date and time representation.	|Tue Aug 16 21:30:00 1988 (en_US); Di 16 Aug 21:30:00 1988 (de_DE)|(1)|
|%x	|Locale’s appropriate date representation.	|08/16/88 (None); 08/16/1988 (en_US); 16.08.1988 (de_DE)|(1)|
|%X	|Locale’s appropriate time representation.	|21:30:00 (en_US); 21:30:00 (de_DE)|(1)|
|%%	|A literal '%' character.	|%	 | |

## Calendar

In [88]:
import calendar

In [89]:
wd = calendar.Calendar()

In [90]:
for d in wd.iterweekdays():
    print(d)

0
1
2
3
4
5
6


In [91]:
for d in wd.itermonthdates(month=10, year=1978):
    print(d)

1978-09-25
1978-09-26
1978-09-27
1978-09-28
1978-09-29
1978-09-30
1978-10-01
1978-10-02
1978-10-03
1978-10-04
1978-10-05
1978-10-06
1978-10-07
1978-10-08
1978-10-09
1978-10-10
1978-10-11
1978-10-12
1978-10-13
1978-10-14
1978-10-15
1978-10-16
1978-10-17
1978-10-18
1978-10-19
1978-10-20
1978-10-21
1978-10-22
1978-10-23
1978-10-24
1978-10-25
1978-10-26
1978-10-27
1978-10-28
1978-10-29
1978-10-30
1978-10-31
1978-11-01
1978-11-02
1978-11-03
1978-11-04
1978-11-05


In [92]:
wd.monthdatescalendar(1978, 10)

[[datetime.date(1978, 9, 25),
  datetime.date(1978, 9, 26),
  datetime.date(1978, 9, 27),
  datetime.date(1978, 9, 28),
  datetime.date(1978, 9, 29),
  datetime.date(1978, 9, 30),
  datetime.date(1978, 10, 1)],
 [datetime.date(1978, 10, 2),
  datetime.date(1978, 10, 3),
  datetime.date(1978, 10, 4),
  datetime.date(1978, 10, 5),
  datetime.date(1978, 10, 6),
  datetime.date(1978, 10, 7),
  datetime.date(1978, 10, 8)],
 [datetime.date(1978, 10, 9),
  datetime.date(1978, 10, 10),
  datetime.date(1978, 10, 11),
  datetime.date(1978, 10, 12),
  datetime.date(1978, 10, 13),
  datetime.date(1978, 10, 14),
  datetime.date(1978, 10, 15)],
 [datetime.date(1978, 10, 16),
  datetime.date(1978, 10, 17),
  datetime.date(1978, 10, 18),
  datetime.date(1978, 10, 19),
  datetime.date(1978, 10, 20),
  datetime.date(1978, 10, 21),
  datetime.date(1978, 10, 22)],
 [datetime.date(1978, 10, 23),
  datetime.date(1978, 10, 24),
  datetime.date(1978, 10, 25),
  datetime.date(1978, 10, 26),
  datetime.date(197

In [93]:
wd.monthdayscalendar(1978, 10)

[[0, 0, 0, 0, 0, 0, 1],
 [2, 3, 4, 5, 6, 7, 8],
 [9, 10, 11, 12, 13, 14, 15],
 [16, 17, 18, 19, 20, 21, 22],
 [23, 24, 25, 26, 27, 28, 29],
 [30, 31, 0, 0, 0, 0, 0]]

In [94]:
dir(wd)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_firstweekday',
 'firstweekday',
 'getfirstweekday',
 'itermonthdates',
 'itermonthdays',
 'itermonthdays2',
 'itermonthdays3',
 'itermonthdays4',
 'iterweekdays',
 'monthdatescalendar',
 'monthdays2calendar',
 'monthdayscalendar',
 'setfirstweekday',
 'yeardatescalendar',
 'yeardays2calendar',
 'yeardayscalendar']